drivers/nrf_wifi/fw_if/umac_if/src/offload_raw_tx/fmac_api.c

```diff
@@ -0,0 +1,312 @@
+				      __func__);
+		status = NRF_WIFI_STATUS_FAIL;
+		goto out;
+	}
+
+	status = NRF_WIFI_STATUS_SUCCESS;
+out:
+	return status;
+}
+
+enum nrf_wifi_status nrf_wifi_fmac_off_raw_tx_dev_init(
+		struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+#ifdef NRF_WIFI_LOW_POWER
+		int sleep_type,
+#endif /* NRF_WIFI_LOW_POWER */
+		unsigned int phy_calib,
+		enum op_band op_band,
+		bool beamforming,
+		struct nrf_wifi_tx_pwr_ctrl_params *tx_pwr_ctrl_params,
+		struct nrf_wifi_tx_pwr_ceil_params *tx_pwr_ceil_params,
+		struct nrf_wifi_board_params *board_params)
+{
+	enum nrf_wifi_status status = NRF_WIFI_STATUS_FAIL;
+	struct nrf_wifi_fmac_otp_info otp_info;
+	struct nrf_wifi_phy_rf_params phy_rf_params;
+
+	if (!fmac_dev_ctx) {
+		nrf_wifi_osal_log_err("%s: Invalid device context",
+				      __func__);
+		goto out;
+	}
+
+	status = nrf_wifi_hal_dev_init(fmac_dev_ctx->hal_dev_ctx);
+	if (status != NRF_WIFI_STATUS_SUCCESS) {
+		nrf_wifi_osal_log_err("%s: nrf_wifi_hal_dev_init failed",
+				      __func__);
+		goto out;
+	}
+
+	fmac_dev_ctx->tx_pwr_ceil_params = nrf_wifi_osal_mem_alloc(sizeof(*tx_pwr_ceil_params));
+	nrf_wifi_osal_mem_cpy(fmac_dev_ctx->tx_pwr_ceil_params,
+			      tx_pwr_ceil_params,
+			      sizeof(*tx_pwr_ceil_params));
+
+	nrf_wifi_osal_mem_set(&otp_info,
+			      0xFF,
+			      sizeof(otp_info));
+
+	status = nrf_wifi_hal_otp_info_get(fmac_dev_ctx->hal_dev_ctx,
+					   &otp_info.info,
+					   &otp_info.flags);
+	if (status != NRF_WIFI_STATUS_SUCCESS) {
+		nrf_wifi_osal_log_err("%s: Fetching of RPU OTP information failed",
+				      __func__);
+		goto out;
+	}
+
+	status = nrf_wifi_fmac_rf_params_get(fmac_dev_ctx,
+		                             &phy_rf_params);
+
+	if (status != NRF_WIFI_STATUS_SUCCESS) {
+		nrf_wifi_osal_log_err("%s: RF parameters get failed",
+				     __func__);
+		goto out;
+	}
+
+	status = nrf_wifi_fmac_off_raw_tx_fw_init(fmac_dev_ctx,
+						  &phy_rf_params,
+						  true,
+#ifdef NRF_WIFI_LOW_POWER
+						  sleep_type,
+#endif /* NRF_WIFI_LOW_POWER */
+						  phy_calib,
+						  op_band,
+						  beamforming,
+						  tx_pwr_ctrl_params,
+						  board_params);
+
+	if (status == NRF_WIFI_STATUS_FAIL) {
+		nrf_wifi_osal_log_err("%s: nrf_wifi_fmac_off_raw_tx_fw_init failed",
+				      __func__);
+		goto out;
+	}
+out:
+	return status;
+}
+
+void nrf_wifi_fmac_off_raw_tx_dev_rem(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx)
+{
+	nrf_wifi_hal_dev_rem(fmac_dev_ctx->hal_dev_ctx);
+	nrf_wifi_osal_mem_free(fmac_dev_ctx);
+}
+
+void nrf_wifi_fmac_off_raw_tx_dev_deinit(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx)
+{
+	nrf_wifi_osal_mem_free(fmac_dev_ctx->tx_pwr_ceil_params);
+}
+
+void nrf_wifi_fmac_off_raw_tx_deinit(struct nrf_wifi_fmac_priv *fpriv)
+{
+	nrf_wifi_hal_deinit(fpriv->hpriv);
+	nrf_wifi_osal_mem_free(fpriv);
+}
+
+static bool valid_rates(enum nrf_wifi_off_raw_tx_tput_mode tput_mode,
+			enum nrf_wifi_off_raw_tx_rate rate)
+{
+	if (tput_mode == TPUT_MODE_LEGACY){
+		if (rate > RATE_54M) {
```

Comment:
Where is `RATE_54M` defined. Note that we cannot use it from `off_raw_tx_api.h` since that is a zephyr specific include file which cannot be included in OSAL code.

Reply:
Done.